common/djangoapps/student/tests/tests.py

```diff
@@ -930,3 +933,221 @@ def test_roundtrip_with_unicode_course_id(self):
         self.assertEqual(self.user, real_user)
         self.assertEqual(anonymous_id, anonymous_id_for_user(self.user, course2.id, save=False))
+
+
+@ddt.ddt
+class DashboardTestXSeriesPrograms(ModuleStoreTestCase, ProgramsApiConfigMixin):
+    """
+    Tests for dashboard for xseries program courses. Enroll student into
+    three programs and then try different combinations to see xseries upsell
+    messages are appearing.
+    """
+    def setUp(self):
+        super(DashboardTestXSeriesPrograms, self).setUp()
+
+        self.user = UserFactory.create(username="jack", email="jack@fake.edx.org", password='test')
+        self.course = CourseFactory.create()
+        self.course_two = CourseFactory.create()
+        self.course_three = CourseFactory.create()
+        self.program_name = 'Testing Program'
+        self.category = 'xseries'
+
+        CourseModeFactory.create(
+            course_id=self.course.id,
+            mode_slug='verified',
+            mode_display_name='Verified',
+            expiration_datetime=datetime.now(pytz.UTC) + timedelta(days=1)
+        )
+        self.client = Client()
+        cache.clear()
+
+    def _create_program_data(self, data):
+        """Dry method to create testing programs data."""
+        programs = {}
+        for course, program_status in data:
+            programs[unicode(course)] = {
+                'category': self.category,
+                'organization': {'display_name': 'Test Organization 1', 'key': 'edX'},
+                'marketing_slug': 'fake-marketing-slug-xseries-1',
+                'status': program_status,
+                'course_codes': [
+                    {
+                        'display_name': 'Demo XSeries Program 1',
+                        'key': 'TEST_A',
+                        'run_modes': [{'sku': '', 'mode_slug': 'ABC_1', 'course_key': unicode(course)}]
+                    },
+                    {
+                        'display_name': 'Demo XSeries Program 1',
+                        'key': 'TEST_A',
+                        'run_modes': [{'sku': '', 'mode_slug': 'ABC_1', 'course_key': unicode(course)}]
+                    },
+                    {
+                        'display_name': 'Demo XSeries Program 1',
+                        'key': 'TEST_A',
+                        'run_modes': [{'sku': '', 'mode_slug': 'ABC_1', 'course_key': unicode(course)}]
+                    }
+                ],
+                'subtitle': 'sub',
+                'name': self.program_name
+            }
+
+        return programs
+
+    @ddt.data(
+        ('active', [{'sku': ''}, {'sku': ''}, {'sku': ''}, {'sku': ''}], 'slug1'),
+        ('active', [{'sku': ''}, {'sku': ''}, {'sku': ''}], 'slug2'),
+        ('active', [], ''),
+        ('unpublished', [{'sku': ''}, {'sku': ''}, {'sku': ''}, {'sku': ''}], 'slug3'),
+    )
+    @ddt.unpack
+    def test_get_xseries_programs_method(self, status, course_codes, slug):
```

Comment:
Add doc string may be something like
`Verify that data is fetched correctly for given course.`